Sources/BSON/BSONDocument.swift

```diff
@@ -174,12 +185,107 @@ public struct BSONDocument {
      */
     public subscript(dynamicMember member: String) -> BSON? {
         get { self[member] }
-        set { fatalError("Unimplemented") }
+        set { self[member] = newValue }
+    }
+
+    /**
+     * Sets a BSON element with the corresponding key
+     * if element.value is nil the element is deleted from the BSON
+     */
+    internal mutating func set(key: String, of value: BSON?) throws {
+        if !self.keySet.contains(key), let value = value {
+            // appending new key
+            self.keySet.insert(key)
+            self._buffer.moveWriterIndex(to: self.byteLength - 1) // setup to overwrite null terminator
+            let size = BSONDocument.append(element: (key, value), to: &self._buffer)
+            self._buffer.writeInteger(0, endianness: .little, as: UInt8.self) // add back in our null terminator
+            self.byteLength += size
+            return
+        }
+
+        guard value != nil || self.keySet.contains(key) else {
+            // no-op: trying to delete a key that doesn't exist
+            return
+        }
+
+        var iter = BSONDocumentIterator(over: self._buffer)
+
+        guard let (start, end) = iter.findByteRange(for: key) else {
+            throw BSONError.InternalError(message: "Cannot find \(key) to delete")
+        }
+
+        var newBuffer = BSON_ALLOCATOR.buffer(capacity: 0)
+
+        guard
+            let prefix = self._buffer.getBytes(at: 0, length: start),
+            let suffix = self._buffer.getBytes(at: end, length: self.byteLength - end)
+        else {
+            throw BSONError.InternalError(
+                message: "Cannot slice buffer from " +
+                    "0 to len \(start) and from \(end) to len \(self.byteLength - end) : \(self._buffer)"
+            )
+        }
+
+        newBuffer.writeBytes(prefix)
+
+        var newSize = self.byteLength - (end - start)
+        if let value = value {
+            // Overwriting
+            let size = BSONDocument.append(element: (key, value), to: &newBuffer)
+            newSize += size
+
+            guard newSize != BSON_MAX_SIZE else {
+                throw BSONError.DocumentTooLargeError(value: value.bsonValue, forKey: key)
+            }
+        } else {
+            // Deleting
+            self.keySet.remove(key)
+        }
+
+        newBuffer.writeBytes(suffix)
+
+        self._buffer = newBuffer
+        self.byteLength = newSize
+        guard self.byteLength == self._buffer.readableBytes else {
+            fatalError("I think the bson is \(self.byteLength) but I can only read \(self._buffer.readableBytes)")
```

Comment:
I think we generally try to avoid speaking in the first person in error messages, can we rephrase?

Reply:
Silly debug phrasing 😅  nice catch